src/gui/PasswordEdit.cpp

```diff
@@ -105,3 +114,45 @@ void PasswordEdit::autocompletePassword(const QString& password)
         setText(password);
     }
 }
+
+/**
+ * Set the status of secure input entry on macOS. This stops keyboard intercept
+ * processes (e.g. keyloggers, accessibility services) from reading keypresses.
+ *
+ * It's important to turn off this off when not needed to avoid interfering with
+ * accessibility functionality and other legitimate uses of keyboard event taps.
+ *
+ * See the Apple Technical Note 2150:
+ * https://developer.apple.com/library/archive/technotes/tn2150/_index.html
+ *
+ * This functionality was adapted from the following Qt 4.8 source files:
+ *     src/gui/kernel/qkeymapper_mac.cpp
+ *     src/gui/widgets/qlineedit.cpp
+ *
+ * @param enabled Status to set secure input entry to
+ */
+void PasswordEdit::secureInputEntry(bool enabled)
+{
+#ifdef Q_OS_MACOS
+    // are we currently in secure input entry mode?
+    static bool secure = false;
```

Comment:
Don't use static variables inside member functions. This variable is shared between ALL INSTANCES of the PasswordEdit class. I would not even worry about the "current state" and just reissue the enable/disable function calls:
```
if (state) {
	EnableSecureEventInput();
} else {
	DisableSecureEventInput();
}
```

Reply:
From [docs](https://web.archive.org/web/20090418151145/http://developer.apple.com/documentation/Carbon/Reference/Carbon_Event_Manager_Ref/Reference/reference.html#//apple_ref/c/func/EnableSecureEventInput):

> This function maintains a count of the number of times that it has been called. Secure event input is not disabled until DisableSecureEventInput has been called the same number of times. Be sure to disable secure event input if your application becomes inactive. If your application crashes, secure event input is automatically disabled if no other application has enabled it.

What's a better way of implementing this?

Reply:
I gave it to you. Just call the Enable/Disable functions no matter what. There is no need to restrict calls to these functions, I am assuming they are no-op if already set.

Reply:
Oh that is a stupid implementation!

Reply:
You should probably monitor the output of this function and loop calling the DisableSecureEventInput() function until this returns false. This should be done when the password field loses focus.

https://web.archive.org/web/20090418151145/http://developer.apple.com/documentation/Carbon/Reference/Carbon_Event_Manager_Ref/Reference/reference.html#//apple_ref/c/func/IsSecureEventInputEnabled

Reply:
I agree. Is there a chance then that a call to `DisableSecureEventInput()` would be missed? Is there a way for an instance of the password edit field being destroyed while it still has focus?

Reply:
> You should probably monitor the output of this function and loop calling the DisableSecureEventInput() function until this returns false.

According to the docs, `IsSecureEventInputEnabled()` will return `true` if any process has enabled secure entry. But obviously we can't override that and disable it. So this kind of loop would run until the other app disables secure entry.

Reply:
I'm beginning to not like this secure event input!

Reply:
I just realized we were looking at old documentation. There is no mention of Carbon or these function calls in recent documentation. Is there a different way to handle this in 10.14+? Perhaps a Cocoa implementation?